File: soonba/boj_1213.java
/*
 * BAEKJOON ONLINE JUDGE
 * https://www.acmicpc.net
 * Problem Number: 1213
 * Level: Silver III
 * Algorithm: String
 */

/*
입력값 분리
솔팅하여 queue 에 담는다.
queue 에서 poll()하며 element()와 일치하지 않을 때까지 개수를 센다.
count 의 홀짝에 따라 string 배열을 채워간다.
count 홀수는 없거나 1개만 존재해야하므로, boolean 으로 판단하여 불가능한 경우를 체크한다.
string 배열을 모두 채우면, 최종 문자열을 출력한다.
 */

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String string = br.readLine();
        String[] split = string.split("");

        List<String> strList = new ArrayList<>(List.of(split));
        Collections.sort(strList);
        Queue<String> queue = new LinkedList<>(strList);

        boolean middleStr = false;
        boolean isBreak = false;

        int len = split.length;
        String[] sbArray = new String[len];
        int totalIndex = 0;
        while(!queue.isEmpty()) {
            String temp = queue.poll();
            int count = 1;
            if(!queue.isEmpty()){
                while(queue.size()-1 >= 0 && queue.element().equals(temp)) {
                    queue.poll();
                    count++;
                }
            }
            int i1 = count / 2;
            for (int i = 0; i < i1; i++) {
                sbArray[totalIndex] = temp;
                sbArray[len-1-totalIndex] = temp;
                totalIndex++;
            }
            if(count%2 ==1) {
                if(middleStr) {
                    isBreak = true;
                    System.out.println("I'm Sorry Hansoo");
                    break;
                }
                sbArray[(len/2)] = temp;
                middleStr = true;
            }
        }
        StringBuilder sb = new StringBuilder();

        if(!isBreak) {
            for (String s : sbArray) {
                sb.append(s);
            }
        }
        System.out.println(sb);
    }

}

File: sykim/boj_1213.js
/*
 * BAEKJOON ONLINE JUDGE
 * https://www.acmicpc.net
 * Problem Number: 1213
 * Level: Silver III
 * Algorithm: String
 */

/* Pseudocode *
문자열을 오름차순으로 정렬한 후 map을 만들어 문자의 개수를 저장한다.
만든 map을 전개하여 아래 규칙에 따라 answer과 middle을 구한다.
  - count가 홀수일 경우 middle에 저장하고 count가 홀수이면서 3이상일 경우 parseInt(count / 2)만큼 repeat하여 answer에 추가한다.
    이 떄, middle에 이미 저장된 값이 존재한다면 팰린드롬을 만들 수 없으므로, isPalindrome을 false로 바꿔준다.
  - cousnt가 짝수라면, count / 2만큼 repeat하여 answer에 추가한다.
isPalindrome이 true이면 answer + middle + answer.split("").sort().reverse().join("") 을 출력하고,
isPalindrome이 false이면 I'm Sorry Hansoo를 출력한다.
*/

const input = require("fs")
  .readFileSync(process.platform === "linux" ? "/dev/stdin" : "example.txt")
  .toString()
  .trim()
  .split("")
  .sort();

const map = new Map();

input.forEach((item) => {
  if (map.has(item)) map.set(item, map.get(item) + 1);
  else map.set(item, 1);
});

let middle = "";
let answer = "";
let isPalindrome = true;

const solve = () => {
  [...map].forEach((item) => {
    const [alphabet, count] = item;
    if (count % 2) {
      if (middle) {
        isPalindrome = false;
      } else {
        middle = alphabet;
        if (count > 1) {
          answer += alphabet.repeat(parseInt(count / 2));
        }
      }
    } else {
      answer += alphabet.repeat(count / 2);
    }
  });

  if (isPalindrome)
    return answer + middle + answer.split("").sort().reverse().join("");
  else return "I'm Sorry Hansoo";
};

console.log(solve());
